graph/path/earliest.go

```diff
@@ -0,0 +1,100 @@
+package path
+
+import (
+	"gonum.org/v1/gonum/graph"
+	"log"
+	"gonum.org/v1/gonum/graph/temporal"
+)
+
+type Earliest struct {
+	from  graph.Node
+	at    uint64
+	until uint64
+	nodes map[int64]struct{
+		earliest uint64
+		via      []int64
+	}
+}
+
+func (e *Earliest) set(v graph.Node, t uint64, p []int64) {
+	e.nodes[v.ID()] = struct{
+		earliest uint64
+		via      []int64
+	}{
+		t,
+		p,
+	}
+}
+
+func (e *Earliest) From() graph.Node { return e.from }
+
+func (e *Earliest) At() uint64 { return e.at }
+
+func (e *Earliest) Until() uint64 { return e.until }
+
+func (e *Earliest) To(u interface{}) (path []graph.Node, duration uint64) {
+	var uid int64
+	switch t := u.(type) {
+	case graph.Node:
+		uid = t.ID()
+	case int64:
+		uid = t
+	default:
+		log.Panicln("not supported")
+	}
+	if eu, ok := e.nodes[uid]; !ok {
+		return nil, temporal.MaxTime
+	} else {
+		duration = eu.earliest
+		for _, vid := range eu.via {
+			v := temporal.Node(vid)
+			path = append(path, &v)
+		}
+		u := temporal.Node(uid)
+		path = append(path, &u)
+		return
+	}
+}
+
+func (e *Earliest) Len() int {
+	return len(e.nodes)
+}
+
+// EarliestArrivalFrom computes the earliest-arrival paths to all nodes reachable from u where
+// from <= t <=until. It is an implementation of Algorithm 1. described in
+// https://www.vldb.org/pvldb/vol7/p721-wu.pdf
+func EarliestArrivalFrom(g graph.TemporalStream, from graph.Node, at uint64, until uint64) Earliest {
```

Comment:
This could just be `EarliestArrivalFrom(g graph.Graph, from graph.Node, at uint64, until uint64) Earliest` and then type switch on the capacity to get a temporal stream, defaulting to [-inf,inf] temporal intervals when unable to do that.

Reply:
What? No.

Even with [-inf,inf] time constraint the edges still need to have start time and otherwise it's not earliest-arrival so much as some sort of weird DFS based on in what order the edges are provided.

Reply:
The paper says that co-occurrent edges can be arbitrarily ordered. That's what this does.

Reply:
No. It is setting the time bounds of the search, which has no effect on the start times of the edges. It only affects the scope of which edges may be traversed.

Reply:
In the case that no temporal information is know, the graph is a `graph.Directed` only, then the algorithm degenerates to the case I am talking about. Sure, it's not necessarily useful, but I'm considering the API here as well. If this were all in temporal, I'd be pushing back less on this.

The gains that the algorithm gives you depend on a specified order. We have other algorithms here that have similar invariants and they either do the work themselves or require that the graph be handed to them that way (this varies). Here we have a time ordering invariant, but the graph itself does not do that work, instead it's done lazily by the `OrderedTemporalLines`. This means that (unless you are planning on changing that behaviour), it could be done initially by `EarliestArrivalFrom` after obtaining the list of edges. There is the possibility that you are considering the use of a non-memory edge store though (that breaks the assumption that we can fit all the edges into memory). If that's the case, then this is reasonable.

Reply:
> In the case that no temporal information is know, the graph is a graph.Directed only, then the algorithm degenerates to the case I am talking about.

Yes, after another look this is clear now.

> Here we have a time ordering invariant, but the graph itself does not do that work, instead it's done lazily by the OrderedTemporalLines

My intention is that the `TemporalStream` implementor will provide an already-ordered iterator for the reason you mentioned.

Having `OrderedTemporalLines` do this is
- for convenience, no point for the user to reimplement this.
- lazily, with a copy, for immutability. The type name suggests the order is important and there is no other means to change it for any of the 'Ordered' types.

Reply:
Yeah, don't copy here.

I've been thinking about this over the day and ISTM that all we need to add is a single interface (`StreamGraph interface { graph.DirectedMultigraph; Stream() graph.Lines }`) in a new package temporal which would also hold the actual functionality, allow with the container types. It's arguable that even an extra iterator is not really required, but that can be explored (`Stream` either returns the new type or has documentation indicating that the `graph.Lines` must return `graph.Line`s that can be type asserted to the `temporal.Line` type).

Reply:
SGTM, however

> a single interface (StreamGraph interface { graph.DirectedMultigraph; Stream() graph.Lines }) in a new package temporal

Maybe it's overkill to require implementation of `graph.DirectedMultigraph` when we already know we can run algorithms only with the `Stream()` representation of the graph.

And for this reason I think that the interface we want is `LineStreamer interface { LineStream() graph.Lines }`

Reply:
That seems fair.